lib/resumption/resumable.go

```diff
@@ -0,0 +1,348 @@
+	var stopRequested atomic.Bool
+	requestStop := func() {
+		if !stopRequested.Swap(true) {
+			r.cond.Broadcast()
+		}
+	}
+	r.requestDetach = func() {
+		nc.Close()
+		requestStop()
+	}
+	r.cond.Broadcast()
+
+	defer func() {
+		r.mu.Lock()
+		defer r.mu.Unlock()
+		r.requestDetach = nil
+		r.cond.Broadcast()
+	}()
+
+	localPosition := r.receiveBuffer.end
+	r.mu.Unlock()
+
+	ncReader, ok := nc.(byteReaderReader)
+	if !ok {
+		ncReader = bufio.NewReader(nc)
+	}
+
+	var peerPosition uint64
+	if !firstConn {
+		p, err := resumeV1Handshake(r, nc, ncReader, localPosition)
+		if err != nil {
+			return trace.Wrap(err, "handshake")
+		}
+		peerPosition = p
+	}
+
+	var eg errgroup.Group
+
+	eg.Go(func() error {
+		defer requestStop()
+		// the read loop exits on I/O errors (which will kill the write loop
+		// too) but also upon receiving an error tag from the remote, signaling
+		// that the peer has already been done with the connection for a while
+		// now, so anything we're going to write is going to be useless anyway
+		defer nc.Close()
+		return trace.Wrap(runResumeV1Read(r, ncReader, &stopRequested), "read loop")
+	})
+
+	eg.Go(func() error {
+		defer requestStop()
+		// we shouldn't close the connection when exiting from the write loop,
+		// because the read loop might have data still worth parsing (if we
+		// exited because of I/O errors)
+		return trace.Wrap(runResumeV1Write(r, nc, &stopRequested, localPosition, peerPosition), "write loop")
+	})
+
+	return trace.Wrap(eg.Wait())
+}
+
+func resumeV1Handshake(r *Conn, nc net.Conn, ncReader byteReaderReader, localPosition uint64) (peerPosition uint64, err error) {
+	handshakeWatchdog := time.AfterFunc(handshakeTimeout, func() { nc.Close() })
+	defer handshakeWatchdog.Stop()
+
+	var eg errgroup.Group
+	eg.Go(func() error {
+		_, err := nc.Write(binary.AppendUvarint(nil, localPosition))
+		return trace.Wrap(err, "writing local receive position")
+	})
+	eg.Go(func() error {
+		var err error
+		peerPosition, err = binary.ReadUvarint(ncReader)
+		return trace.Wrap(err, "reading peer receive position")
+	})
+	if err := eg.Wait(); err != nil {
+		return 0, trace.Wrap(err)
+	}
+
+	r.mu.Lock()
+	if minPos, maxPos := r.sendBuffer.start, r.sendBuffer.end; peerPosition < minPos || maxPos < peerPosition {
+		// incompatible resume position, mark as remotely closed since we can't
+		// ever continue from this; this also includes receiving an errorTag
+		// (since that's too big of a position to reach legitimately)
+		r.remoteClosed = true
+		r.cond.Broadcast()
+		r.mu.Unlock()
+
+		_, _ = nc.Write([]byte(errorTagUvarint))
+		return 0, trace.BadParameter("got incompatible resume position (%v, expected %v to %v)", peerPosition, minPos, maxPos)
+	}
+
+	if r.sendBuffer.start != peerPosition {
+		r.sendBuffer.advance(peerPosition - r.sendBuffer.start)
+		r.cond.Broadcast()
+	}
+	r.mu.Unlock()
+
+	return peerPosition, nil
+}
+
+func runResumeV1Read(r *Conn, nc byteReaderReader, stopRequested *atomic.Bool) error {
```

Comment:
This and `runResumeV1Write` would both benefit a _lot_ from inline comments.  These functions are really really dense and a lot of what they are doing is a bit opaque.  Detailed comments would go a long way toward making these functions more reviewable/maintainable.

Reply:
added